sqsworkers/crew.py

```diff
@@ -1,327 +1,200 @@
-                new_worker = Worker(self.crew)
-                self.crew.logger.info('hired a new worker %s, staring it up...' % (new_worker.worker_name))
-                new_worker.start()
-                self.crew.logger.info('started a new worker %s' % (new_worker.worker_name))
-                good_workers.append(new_worker)
-        self.crew.logger.info('total number of workers after supervision: %s' % (str(len(good_workers))))
-        self.crew.statsd.gauge('workers.employed', len(good_workers), tags=[self.crew.name])
-        self.crew.workers = good_workers
+
+                messages = self.queue.receive_messages(
+                    AttributeNames=["All"],
+                    MessageAttributeNames=["All"],
+                    MaxNumberOfMessages=self.max_number_of_messages,
+                    WaitTimeSeconds=self.wait_time,
+                )
+
+            if messages:
+
+                metadata = [MessageMetadata(m) for m in messages]
+
+                self.logger.info(
+                    f"processing the following {len(messages)} messages in bulk: {metadata}"
+                )
+
+                task: futures.Future = self._executor.submit(
+                    self.message_processor(messages).start
+                )
+
+                task.add_done_callback(
+                    partial(
+                        self._task_complete,
+                        messages=messages,
+                        metadata=metadata,
+                    )
+                )
+
+                self.statsd.increment(
+                    "process.record.start", len(messages), tags=[]
+                )
+
+            time.sleep(self.polling_interval)
+
+    def _task_complete(
+        self,
+        f: futures.Future,
+        messages: Iterable[Any],
+        metadata=Optional[List[MessageMetadata]],
+    ):
+        """Clean up after task and do any necessary logging."""
+
+        messages = list(messages)
+
+        metadata: List[MessageMetadata] = [
+            MessageMetadata(m) for m in messages
+        ] if metadata is None else metadata
+
+        exception = f.exception()
+
+        if exception is not None:
+
+            self.logger.error(
+                "failed processing {messages} with the following exception: {exception}".format(
+                    exception=repr(exception), messages=metadata
+                ),
+                exc_info=exception,
+            )
+
+            self.statsd.increment(
+                "process.record.failure", len(messages), tags=[]
+            )
+
+        else:
+
+            self.logger.info(
+                "successfully processed {messages}".format(messages=metadata)
+            )
+
+            self.statsd.increment(
+                "process.record.success", len(messages), tags=[]
+            )
+
+            # make sure we don't try to delete more than 10 messages
+            # at a time or we'll get an error from boto3
+            while messages:
```

Comment:
There could be the case where we pulled 10 messages off, 8 were processed successfully, but 2 were not so we should delete 8 but not the two.

Reply:
same comment as above